reward/addressBookManager.go

```diff
@@ -0,0 +1,198 @@
+		return
+	}
+
+	if len(*allTypeList) != len(*allAddressList) {
+		err = errors.New(fmt.Sprintf("length of type list and address list differ. len(type)=%d, len(addrs)=%d", len(*allTypeList), len(*allAddressList)))
+		return
+	}
+
+	// Parse and construct node information
+	for i, addrType := range *allTypeList {
+		switch addrType {
+		case addressTypeNodeID:
+			nodeIds = append(nodeIds, (*allAddressList)[i])
+		case addressTypeStakingAddr:
+			stakingAddrs = append(stakingAddrs, (*allAddressList)[i])
+		case addressTypeRewardAddr:
+			rewardAddrs = append(rewardAddrs, (*allAddressList)[i])
+		case addressTypePoCAddr:
+			pocAddr = (*allAddressList)[i]
+		case addressTypeKIRAddr:
+			kirAddr = (*allAddressList)[i]
+		default:
+			err = errors.New(fmt.Sprintf("invalid type from AddressBook: %d", addrType))
+			return
+		}
+	}
+
+	// validate parsed node information
+	if len(nodeIds) != len(stakingAddrs) ||
+		len(nodeIds) != len(rewardAddrs) ||
+		isEmptyAddress(pocAddr) ||
+		isEmptyAddress(kirAddr) {
+		err = errAddressBookIncomplete
+		return
+	}
+
+	return
+}
+
+// getStakingInfoFromAddressBook returns staking info when calling AddressBook
+// succeeded. It returns an error otherwise.
+func (abm *addressBookManager) getStakingInfoFromAddressBook(blockNum uint64) (*StakingInfo, error) {
+	if !params.IsStakingUpdateInterval(blockNum) {
+		return nil, errors.New(fmt.Sprintf("not staking block number. blockNum: %d", blockNum))
+	}
+
+	// Prepare a message
+	msg, err := abm.makeMsgToAddressBook()
+	if err != nil {
+		return nil, errors.New(fmt.Sprintf("failed to make message for AddressBook. root err: %s", err))
+	}
+
+	intervalBlock := abm.bc.GetBlockByNumber(blockNum)
+	if intervalBlock == nil {
+		return nil, errors.New("stateDB is not ready for staking info")
+	}
+	statedb, err := abm.bc.StateAt(intervalBlock.Root())
+	if err != nil {
+		return nil, errors.New(fmt.Sprintf("failed to make a state for interval block. blockNum: %d, root err: %s", blockNum, err))
+	}
+
+	// Create a new context to be used in the EVM environment
+	context := blockchain.NewEVMContext(msg, intervalBlock.Header(), abm.bc, nil)
+	evm := vm.NewEVM(context, statedb, abm.bc.Config(), &vm.Config{})
+
+	res, gas, kerr := blockchain.ApplyMessage(evm, msg)
+	logger.Trace("Call AddressBook contract", "used gas", gas, "kerr", kerr)
+	err = kerr.ErrTxInvalid
+	if err != nil {
+		return nil, errors.New(fmt.Sprintf("failed to call AddressBook contract. root err: %s", err))
+	}
+
+	nodeIds, stakingAddrs, rewardAddrs, PoCAddr, KIRAddr, err := abm.getAllAddressFromAddressBook(res)
+	if err != nil {
+		if err == errAddressBookIncomplete {
+			// This is expected behavior when addressBook contract is not setup yet.
+			logger.Info("Use empty staking info instead of info from AddressBook", "reason", err)
+		} else {
+			logger.Error("Failed to parse result from AddressBook contract. Use empty staking info", "err", err)
```

Comment:
How about this since we cannot know the exact type of the error.

`logger.Error("Use empty staking info", "err", err)`

Reply:
This is not an error case. It is executed if addressBook contract is not activated.

Reply:
Discussed f2f.